lib/Net/Amazon/S3/Signature/V4Implementation.pm

```diff
@@ -58,13 +59,12 @@ Note that the access key ID is an alphanumeric string, not your account ID. The
 
 sub new {
 	my $class = shift;
-	my ( $access_key_id, $secret, $endpoint, $service ) = @_;
-	my $self = {
-		access_key_id => $access_key_id,
-		secret        => $secret,
-		endpoint      => $endpoint,
-		service       => $service,
-	};
+	my $self = {};
+	if (int(@_) > 1) {
```

Comment:
Please, take a look at `Sub::Params`

Reply:
Thanks for the comment. I haven't known this module. I still think that increasing the number of dependencies is not necessary in this case. Is this a blocker for you?

Reply:
I'm fun of expressing intention, using this module to express that is imho far better way than just implementing it. In corporate review that will be blocker for me.

Anyway, as far as Net-Amazon-Signature-V4#5 is still not merged I'll vote for removing positional arguments in this implementation at all.

Anyway, putting security token in Signature implementation is imho just hotfix to make it work. I'd like to make protocol (as much as possible) not strictly tied with Amazon implementation (ie moving IAM support with `VM::EC2::Security::CredentialCache` dependency into another pluggable module)

I'd prefer to create `Net::Amazon::S3::Authorization` (and `::IAM`) class(es) with interface:
- `access_key`
- `secret_key`
- `authorization_headers` (returning `x-amz-security-token => ...` hash in '::IAM' case)
- `endpoint` and `service` either methods or available via generic `context` method

`Net::Amazon::S3` will get
- `authorization_class` property (default from `our $DEFAULT_AUTHORIZATION_CLASS` so it can be modified by other module)
- `authorization_context` (built during `BUILD`)

Is it too much to ask you for such refactoring ?

Motivation is to allow to use this library with any authorization scheme, either used by another S3 authorization or future Amazon changes, with usage like this
```
use Net::Amazon::S3;
use Net::Amazon::S3::Use::IAM;
```